ocaml-lsp-server/src/custom_requests/req_merlin_call_compatible.ml

```diff
@@ -0,0 +1,156 @@
+  | Some uri -> Some (Uri.t_of_yojson uri)
+  | _ ->
+    (* If the parameter is incorrectly formatted or missing, we refuse to build
+       the parameter, [uri] is mandatory. *)
+    None
+
+let partition_args =
+  (* The function is relatively optimistic and attempts to treat literal data as
+     strings of characters. *)
+  function
+  | `String s -> Either.Left s
+  | `Bool b -> Either.Left (string_of_bool b)
+  | `Float f -> Either.Left (string_of_float f)
+  | `Int i -> Either.Left (string_of_int i)
+  | `Intlit i -> Either.Left i
+  | x -> Either.Right x
+
+let args_from_structured_json params =
+  match List.assoc_opt "args" params with
+  | Some (`List args) -> (
+    (* Only lists containing stringish data are accepted. *)
+    match List.partition_map ~f:partition_args args with
+    | args, [] -> Some args (* No invalid arg. *)
+    | _, _ :: _ -> None (* Some invalid args. *))
+  | _ ->
+    (* If args is not a list or is absent, we assume that an empty list is
+       passed. *)
+    Some []
+
+let from_structured_json = function
+  | `Assoc params ->
+    let result_as_sexp = as_sexp_from_structured_json params in
+    let open Option.O in
+    let* command = command_from_structured_json params in
+    let* args = args_from_structured_json params in
+    let* uri = uri_from_structured_json params in
+    Some { result_as_sexp; command; args; uri }
+  | _ -> None
+
+let raise_invalid_params ?data ~message () =
+  let open Jsonrpc.Response.Error in
+  raise @@ make ?data ~code:Code.InvalidParams ~message ()
+
+let from_structured_json_exn = function
+  | None -> raise_invalid_params ~message:"Expected params but received none" ()
+  | Some params -> (
+    match from_structured_json params with
+    | Some params -> params
+    | None ->
+      let data =
+        `Assoc
+          [ ("expectedParams", expected_params)
+          ; ("receivedParams", (params :> Json.t))
+          ]
+      in
+      raise_invalid_params ~data ~message:"Unexpected params format" ())
+
+let with_pipeline state uri f =
+  match Document_store.get_opt state.State.store uri with
+  | Some doc ->
+    let merlin = Document.merlin_exn doc in
+    Document.Merlin.with_pipeline_exn merlin f
+  | None ->
+    let message =
+      "Document " ^ Uri.to_string uri ^ " wasn't found in the document store"
+    in
+    raise_invalid_params ~message ()
+
+let parse_arguments state uri specs raw_args cmd_args =
+  let conf = Merlin_config.DB.get state.State.merlin_config uri in
+  let open Fiber.O in
+  let+ config = Merlin_config.config conf in
+  let specs = List.map ~f:snd specs in
+  Mconfig.parse_arguments
+    ~wd:(Sys.getcwd ())
+    ~warning:ignore
+    specs
+    raw_args
+    config
+    cmd_args
+  |> snd
+
+let perform_query action params pipeline =
+  let action () = action pipeline params in
+  let class_, output =
+    match action () with
+    | result -> ("return", result)
+    | exception Failure message -> ("failure", `String message)
```

Comment:
When is this exception possible? if this is a user facing error, it should have its own exception type.